_posts/2018-03-15-integrating-redux-and-socketio.markdown

```diff
@@ -0,0 +1,87 @@
+
+![/images/2018-03-15-integrating-redux-and-socketio/example.gif](/images/2018-03-15-integrating-redux-and-socketio/example.gif)
+
+For server/client, I ended up using socket.io, a battle tested library for two-way real time communication channel. At that point, we were facing another challenge, how do we integrate socket.io in the react/redux architecture we just designed. also written in javascript for backend and clients combine the two. All that’s left to do is to integrate the two libraries.
+
+Here's a simplified version of the helper function that I wrote to broadcast redux actions over the socket to other connected clients via our backend service.
+
+```javascript
+// From client
+import io from 'socket.io-client'
+import { messageTypes } from './messageTypes'
+
+let socket = io(rootURL)
+
+// Helper to emit a redux action to our websocket server
+const emitAction = (actionCreator) => {
+  return (...args) => {
+    // This return the action object which gets sent to our backend
+    // server via the socket connection
+    const result = actionCreator.apply(this, args)
+    socket.emit(result.key, {
+      ...result.payload,
+      type: result.type
+    })
+    return result
+  }
+}
+```
+
+Here's an example of how it's being used to wrap a typical redux action creator. All that's needed is to wrap the function as shown in the following example.
+
+```javascript
+import keyMirror from 'client/lib/keyMirror'
+import { emitAction } from 'client/apps/websocket/client'
+import { messageTypes } from 'client/apps/websocket/messageTypes'
+
+export const actions = keyMirror(
+  ...
+  'START_EDITING_ARTICLE',
+  'STOP_EDITING_ARTICLE',
+  ...
+)
+
+export const startEditingArticle = emitAction((data) => {
+  return {
+    type: actions.START_EDITING_ARTICLE,
+    key: messageTypes.userStartedEditing,
+    payload: {
+      timestamp: new Date().toISOString(),
+      ...data
+    }
+  }
+})
+
+...
+```
+
+After setting it up, all we needed to do is wrap action creators with `emitAction`, add a `key` property to the action object and we can decide to either process actions
+ on the backend or proxy them to other clients. The remainder of the server implementation and the (server handlers)[] is available to inspect [here]() in our github repo. Which has instructions on how to run.
+
+## Future improvements
+
+There's an opportunity to extract this module for other projects and apps, also adding more helpers for handling events in other clients.
+
+## Useful links
+
+- [Redux]()
+- [Socket.io](https://socket.io/)
```

Comment:
solid start 👍

Reply:
thanks 🙏🏾I'll let you know when it's ready for a proper read :)